include/secp256k1_ellswift.h

```diff
@@ -0,0 +1,186 @@
+ */
+
+/** A pointer to a function used by secp256k1_ellswift_xdh to hashing the shared
+ *  X coordinate along with the encoded public keys to a uniform shared secret.
+ *
+ *  Returns: 1 if a shared secret was successfully computed.
+ *           0 will cause secp256k1_ellswift_xdh to fail and return 0.
+ *           Other return values are not allowed, and the behaviour of
+ *           secp256k1_ellswift_xdh is undefined for other return values.
+ *  Out:     output:     pointer to an array to be filled by the function
+ *  In:      x32:        pointer to the 32-byte serialized X coordinate
+ *                       of the resulting shared point (will not be NULL)
+ *           ours64:     pointer to the 64-byte encoded public key we sent
+ *                       to the other party (will not be NULL)
+ *           theirs64:   pointer to the 64-byte encoded public key we received
+ *                       from the other party (will not be NULL)
+ *           data:       arbitrary data pointer that is passed through
+ */
+typedef int (*secp256k1_ellswift_xdh_hash_function)(
+  unsigned char *output,
+  const unsigned char *x32,
+  const unsigned char *ours64,
+  const unsigned char *theirs64,
+  void *data
+);
+
+/** An implementation of an secp256k1_ellswift_xdh_hash_function which uses
+ *  H_tag(key1 || key2 || x32), where (key1, key2) = sort([ours64, theirs64]),
+ *  and ignores data. The sorting is lexicographic. H_tag is the BIP340 tagged
+ *  hash function whose tag is the zero-terminated C string pointed to by data. */
+SECP256K1_API_VAR const secp256k1_ellswift_xdh_hash_function secp256k1_ellswift_xdh_hash_function_tagged;
+
+/** An implementation of an secp256k1_ellswift_xdh_hash_function compatible with
+ *  BIP324. It returns H_tag(init_pubkey || resp_pubkey || x32), where H_tag is
+ *  the BIP340 tagged hash function with tag "bip324_ellswift_xonly_ecdh",
+ *  and init_pubkey is ours64 if *(int*)data is true and theirs64 otherwise.
+ *  resp_pubkey is the other key. */
+SECP256K1_API_VAR const secp256k1_ellswift_xdh_hash_function secp256k1_ellswift_xdh_hash_function_bip324;
+
+/** Construct a 64-byte ElligatorSwift encoding of a given pubkey.
+ *
+ *  Returns: 1 when pubkey is valid.
+ *  Args:    ctx:        pointer to a context object
+ *  Out:     ell64:      pointer to a 64-byte array to be filled
+ *  In:      pubkey:     a pointer to a secp256k1_pubkey containing an
+ *                       initialized public key
+ *           rnd32:      pointer to 32 bytes of randomness
+ *
+ * It is recommended that rnd32 consists of 32 uniformly random bytes, not
+ * known to any adversary trying to detect whether public keys are being
+ * encoded, though 16 bytes of randomness (padded to an array of 32 bytes,
+ * e.g., with zeros) suffice to make the result indistinguishable from
+ * uniform. The randomness in rnd32 must be independent of pubkey. In
+ * particular, it must not be derived deterministically from pubkey.
```

Comment:
nit: this confused me because `must` and `must not` sound like necessary requirements. It's fine to use some hardened derivation of the secret key as rnd32 (which is not independent of the pubkey) and it's fine to derive rnd32 deterministically from pubkey and some secret data.

Reply:
:+1: